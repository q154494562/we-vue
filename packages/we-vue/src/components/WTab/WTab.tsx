// Utils
import mixins, { ExtractVue } from '@/utils/mixins'

// Mixins
import findParent from '@/mixins/find-parent'
import Vue, { VNode } from 'vue'

interface options extends Vue {
  parent: any
}

export default mixins<options &
  ExtractVue<typeof findParent>
>(
  findParent,
).extend({
  name: 'w-tab',

  props: {
    title: String,
    disabled: Boolean,
  },

  data () {
    return {
      inited: false,
    }
  },

  computed: {
    index (): number {
      return this.parent.tabs.indexOf(this) >= 0 ? this.parent.tabs.indexOf(this) : 0
    },

    selected (): boolean {
      return this.index === this.parent.currentActive
    },
  },

  created () {
    this.findParent('w-tabs')
  },

  mounted () {
    const { tabs } = this.parent
    const index = this.parent.$slots.default.indexOf(this.$vnode)
    tabs.splice(index === -1 ? tabs.length : index, 0, this)

    if (this.$slots.title) {
      this.parent.renderTitle(this.$refs.title, this.index)
    }
  },

  beforeDestroy () {
    this.parent.tabs.splice(this.index, 1)
  },

  watch: {
    'parent.currentActive' () {
      this.inited = this.inited || this.selected
    },

    title () {
      this.parent.setLine()
    },
  },

  render (): VNode {
    return (
      <div class="wv-tab__pane" vShow={this.selected || this.parent.animated}>
        { this.inited && this.$slots.default }
        {
          this.$slots.title &&
            <div ref="title">
              {this.$slots.title}
            </div>
        }
      </div>
    )
  },
})
