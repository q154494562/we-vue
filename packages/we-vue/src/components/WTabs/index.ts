import WTabs from './WTabs'

export { WTabs }
export default WTabs
