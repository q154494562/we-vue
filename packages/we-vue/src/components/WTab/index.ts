import WTab from './WTab'

export { WTab }
export default WTab
