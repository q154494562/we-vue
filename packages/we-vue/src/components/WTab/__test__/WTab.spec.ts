import Vue from 'vue'
import { mount } from '@vue/test-utils'
import WTabs from '../../WTabs'
import WTab from '../'

describe('tab', () => {
  test('create', () => {
    const wrapper = mount(WTabs, {
      attachToDocument: true,
      propsData: {},
      slots: {
        default: [WTab, WTab, WTab],
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('destroy', () => {
    const wrapper = mount(WTabs, {
      attachToDocument: true,
      propsData: {},
      slots: {
        default: [WTab, WTab, WTab],
      },
    })

    wrapper.findAll(WTab).at(0).vm.$destroy()

    expect(wrapper.vm.tabs).toHaveLength(2)
  })

  test('render title slot', () => {
    Vue.component('w-tab', WTab)

    const wrapper = mount(WTabs, {
      attachToDocument: true,
      propsData: {},
      slots: {
        default: ['<w-tab><span slot="title">title</span>content1</w-tab>'],
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
