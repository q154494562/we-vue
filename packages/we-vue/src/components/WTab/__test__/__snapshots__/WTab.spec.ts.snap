// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tab create 1`] = `
<div class="wv-tabs wv-tabs--line">
  <div class="wv-tabs__wrap wv-hairline--top-bottom">
    <div class="wv-tabs__nav wv-tabs__nav--line">
      <div class="wv-tabs__line"></div>
    </div>
  </div>
  <div class="wv-tabs__content">
    <div class="wv-tab__pane"></div>
    <div class="wv-tab__pane"></div>
    <div class="wv-tab__pane"></div>
  </div>
</div>
`;

exports[`tab render title slot 1`] = `
<div class="wv-tabs wv-tabs--line">
  <div class="wv-tabs__wrap wv-hairline--top-bottom">
    <div class="wv-tabs__nav wv-tabs__nav--line">
      <div class="wv-tabs__line"></div>
      <div class="wv-tab wv-tab--active"><span class="wv-ellipsis"></span></div>
    </div>
  </div>
  <div class="wv-tabs__content">
    <div class="wv-tab__pane">content1<div><span>title</span></div>
    </div>
  </div>
</div>
`;
